source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -142,146 +142,183 @@ ProcessorState::getCallbackStateAfterHeaderResp(const CommonResponse& common_res
-        // Trailers came in while we were waiting for this response, and the server
-        // is not interested in the body, so send them now.
-        filter_.sendTrailers(*this, *trailers_);
-        clearWatermark();
-        filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
-        return absl::OkStatus();
-      }
+      modifyBufferedData([&common_response](Buffer::Instance& buf) {
+        MutationUtils::applyBodyMutations(common_response.body_mutation(), buf);
+      });
     }
+  }
+
+  // In case any data left over in the chunk queue, clear them.
+  clearStreamingChunk();
+  // Once this message is received, we won't send anything more on this request
+  // or response to the processor. Clear flags to make sure.
+  body_mode_ = ProcessingMode::NONE;
+  send_trailers_ = false;
+  clearWatermark();
 
-    // If we got here, then the processor doesn't care about the body or is not ready for
-    // trailers, so we can just continue.
-    ENVOY_STREAM_LOG(trace, "Clearing stored headers", *filter_callbacks_);
-    headers_ = nullptr;
+  // If we got here, then the processor doesn't care about the body or is not ready for
+  // trailers, so we can just continue.
+  ENVOY_STREAM_LOG(trace, "Clearing stored headers", *filter_callbacks_);
+  headers_ = nullptr;
+  filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+  continueIfNecessary();
+  clearWatermark();
+  return absl::OkStatus();
+}
+
+absl::Status ProcessorState::handleHeaderContinue(const HeadersResponse& response) {
+  if (no_body_) {
+    // Fall through if there was never a body in the first place.
+    ENVOY_STREAM_LOG(debug, "The message had no body", *filter_callbacks_);
+  } else if (complete_body_available_ && body_mode_ != ProcessingMode::NONE) {
+    return handleCompleteBodyAvailable(response);
+  } else if (body_mode_ == ProcessingMode::BUFFERED) {
+    // Here, we're not ready to continue processing because then
+    // we won't be able to modify the headers any more, so do nothing and
+    // let the doData callback handle body chunks until the end is reached.
+    clearWatermark();
+    filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+    return absl::OkStatus();
+  } else if (body_mode_ == ProcessingMode::STREAMED) {
+    sendBufferedDataInStreamedMode(false);
+    filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+    continueIfNecessary();
+    return absl::OkStatus();
+  } else if (body_mode_ == ProcessingMode::FULL_DUPLEX_STREAMED) {
+    // There is no buffered data in this mode.
     filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
     continueIfNecessary();
+    return absl::OkStatus();
+  } else if (body_mode_ == ProcessingMode::BUFFERED_PARTIAL) {
+    return handleBufferedPartialMode(response);
+  }
+
+  return handleTrailersAndCleanup(response);
```

Comment:
Only  if (no_body_) condition will call this return. Thus moving it to that block, i.e, right after line 218 is cleaner.